server/src-lib/Hasura/RQL/DDL/Relationship/Types.hs

```diff
@@ -93,13 +113,27 @@ type ArrRelUsing = RelUsing ArrRelUsingFKeyOn ArrRelManualConfig
 type ArrRelDef = RelDef ArrRelUsing
 type CreateArrRel = WithTable ArrRelDef
 
-newtype ObjRelManualConfig =
-  ObjRelManualConfig { getObjRelMapping :: RelManualConfig }
-  deriving (Show, Eq, FromJSON, ToJSON, Lift)
+data ObjRelUsingFKeyOn
```

Comment:
This type name does not follow the proposed [Haskell style guide](https://github.com/hasura/graphql-engine/blob/950d3842e4662c2ff5d8d5b20ac64bf9f9387b80/server/CONTRIBUTING.md#naming). The name `ObjectRelationshipUsingForeignKeyOn` would be a little long, but maybe there’s a shorter name that could capture what it does without needing to abbreviate?

Taking a step back for a moment, the Ruby on Rails nomenclature for this is that there are three kinds of relationships: `has_one`, `has_many`, and `belongs_to`. The first two describe one-to-one and one-to-many relationships where the foreign table contains the reference, while `belongs_to` describes a one-to-one relationship where the “self” table contains the reference.

Could we have the same kind of explicit modeling? Instead of “array relationship” and “object relationship,” could we have

```haskell
data ForeignColumn = ForeignColumn !QualifiedTable !PGCol

data Relationship
  = HasOne !ForeignColumn
  | HasMany !ForeignColumn
  | BelongsTo !QualifiedTable !PGCol
```

To me, that modeling is a lot more natural than the “object” vs “array” split we have currently, which focuses on the GraphQL interface rather than the actual relational modeling.

Reply:
I agree, I'm not a fan of our current relationship terminology (object/array). I like the explicit nature of rails [associations](https://guides.rubyonrails.org/association_basics.html), there are six kinds as opposed to the two that we have, `has_many` and `belongs_to`  (this PR introduces the third one, `has_one`). 

I'm not sure what's the best way forward though. I would like us to get to rails like simplicity for defining relationships but without disrupting how we treat relationships across our entire codebase. We could for example easily get rid of `insertion_order` and add something like `kind` to specify either `belongs_to` or `has_one`? The end goal could be to have an api like `create_relationship` which takes an extra `kind` argument? The exported metadata could then have `relationships` instead of `object_relationships` and `array_relationships` like we currently have.